### PR TITLE
Whitespaces in paths break for loop

### PR DESCRIPTION
**Pull request by gabyx**
I searched for all `find` statements and found some which do not work with 
whitespaces in path names, I corrected them.

Thanks for reviewing this.

### CODE REVIEW

**Comment by gabyx** [comment]
Jeah, of course.

**Comment by gabyx** [comment]
Maybe lets add the tests later in another PR. We should actually run all tests against some whitespaces. I can probably test this by simply hacking the exec-tests.sh and looking for failed tests.

**Comment by rycus86** [comment]
I'm more concerned about `IFS` on all platforms, so a duplicate test would have the benefit of testing it on all the platforms that are configured already.

**Comment by gabyx** [comment]
Ah men: Shitty IFS crap is so dangerous:
```shell
   echo "$LOCAL_REPOSITORY_LIST" | while IFS= read -r EXISTING; do
        echo "Ex: $EXISTING"
        EXISTING_REPO_ROOT=$(dirname "$EXISTING")
        install_hooks_into_repo "$EXISTING_REPO_ROOT"
    done

```

**Output**
```shell
Do you want to install the hooks into existing repositories? [y/N] Where do you want to start the search? [/root]
LIST: /tmp/test045/001/.git
/tmp/test045/002/.git
with xargs:
f: /tmp/test045/001/.git
f: /tmp/test045/002/.git
Ex: /tmp/test045/001/.git  <<<<.---- WHY ONLY THE FIRST one??
Looks like you don't have a .githooks folder in the /tmp/test045/001 repository yet.
```
does not work, but if I comment out `install_hooks_into_repo` then the loop processes all `\n` delimited paths correctly, otherwise the 

I prefere something more robust instead: `xargs`

**Comment by gabyx** [comment]
Also the decision to not have `'` quotes doesnt enlighten me much, since sometimes you want to control or restrict expansion, and this decision only makes it harder... at least for me...

**Comment by gabyx** [comment]
Ah its because the install script mocks out `</dev/tty` which in my case reads my loop stdin... shit.

**Comment by gabyx** [comment]
The cleanest way I found is the following in f9a214456
the while loop changes the stdinput which is bad and undefined behavior for successive calls.
I testet various methods (for loop, with IFS resetting, looks totally ugly) etc...

**Comment by gabyx** [comment]
I added the test: `step-045w.sh` with whitespaces which would fail if this bugfix is not applied.

**Comment by gabyx** [comment]
I replaced all tests to have whitespaces in a lot of paths (next two commits), this is definitely something we need to check and detect failure early, Thats why I suggest we use all these new tests (`test\d+` -> `test $1`)
They all pass and work 👍 . 
Thanks for reviewing this and merging this in.

File: install.sh [hunk 1]
```diff
@@ -3545,10 +3552,16 @@ install_into_existing_repositories() {
...
     fi
 
-    for EXISTING in $LOCAL_REPOSITORY_LIST; do
+    # We dont want to change the std input through
+    # a while loop, so we redirect over descriptor 3
+    INPUT_PIPE=$(mktemp)
+    rm -rf "$INPUT_PIPE" >/dev/null 2>&1 || true
+    mkfifo "$INPUT_PIPE"
+    echo "$LOCAL_REPOSITORY_LIST" >"$INPUT_PIPE" &
+    while IFS= read -r EXISTING <&3; do
```
Comment (gabyx) [inline install.sh]: Thats the best I could come up with which is POSIX...
If you have smoother way tell me, 
```shell
IFS="
"
for A in $A; do
   cmd-which-uses-std-in
done;
unset IFS
```
Does not work...

File: cli.sh [hunk 2]
```diff
@@ -1147,7 +1147,8 @@ update_shared_hooks_in() {
 
     # split on comma and newline
     IFS=",
-    "
+"
```
Comment (gabyx) [inline cli.sh]: This is super crucial, as it splits also for whitespaces...

**Comment by rycus86** [comment]
Sorry, I was unavailable for a few days, just looking at this now.
82 changed files though doesn't really help. :)
Can we first do the actual change and perhaps add one or two new tests rather than change all the existing ones?

On the actual whitespace splitting problem, I just tried this in `alpine`'s shell:
```
# find . -type f
./def ghi/test
./abc/test file
./jkl/sample
./jkl/mno/another file
# RLIST=$(find . -type f)
#
# this below is what you're trying to fix, right?
# for T in $RLIST; do echo "Item: $T"; done
Item: ./def
Item: ghi/test
Item: ./abc/test
Item: file
Item: ./jkl/sample
Item: ./jkl/mno/another
Item: file
# IFS=",
"
# for T in $RLIST; do echo "Item: $T"; done
Item: ./def ghi/test
Item: ./abc/test file
Item: ./jkl/sample
Item: ./jkl/mno/another file
```

Should we then just ensure all unquoted for loops are executed with the appropriate `IFS` set?

**Comment by gabyx** [comment]
Correct. But setting the IFS is really a bit buggy. Because if we set it, then all commands inside the for loop have the same IFS, which is buggy (when we change things..)
I force pushed two commits, the Bugfix, and all tests with whitespaces...
Could you look at the changes first?
The tests are only whitespace replacements, nothing changed there.

**Comment by gabyx** [comment]
The robust reference is: https://unix.stackexchange.com/questions/9496/looping-through-files-with-spaces-in-the-names

**Comment by gabyx** [comment]
by the way:
image
 In this way I used:
```shell
 # We dont want to change the std input through
    # a while loop, so we redirect over descriptor 3
    INPUT_PIPE=$(mktemp)
    rm -rf "$INPUT_PIPE" >/dev/null 2>&1 || true
    mkfifo "$INPUT_PIPE"
    find "$START_DIR" -path "*templates/hooks/pre-commit.sample*" 2>/dev/null >"$INPUT_PIPE" &
    while IFS= read -r HIT <&3; do
```

**Comment by gabyx** [comment]
I make these changes in the urge to use this tool on windows, but as you know windows is crap and also whitespaces in paths are used a lot there... Maybe we should also add an appveyor test or something like that :-), basically just installing gitforwindows and then run everything there :-)

**Comment by rycus86** [comment]
Thanks! I had the ~~pain~~ privilege of working with Windows for a long time, I don't envy you... :)
For `SC2044`, my understanding is that `find | grep` is less buggy in a for loop than plain `find`s, not sure where I've got this from and could be wrong.
All the changes look OK to me, except the ones that now use temporary files and pipes and whatnot. Is there really nothing simpler we can do? Those are two loops only, even setting `IFS` multiple times would be simpler I think.

**Comment by gabyx** [comment]
Jeah I dont like the pipe also not. Maybe we can just use the 
IFS=
and stick with if it works. Let me change it.

File: cli.sh (hunk 2 above)
Comment (rycus86) [reply cli.sh]: Nice one, good catch!

File: install.sh [hunk 3]
```diff
@@ -3388,7 +3387,11 @@ search_for_templates_dir() {
 search_pre_commit_sample_file() {
     START_DIR="$1"
 
-    for HIT in $(find "$START_DIR" 2>/dev/null | grep "templates/hooks/pre-commit.sample"); do
+    IFS="
+"
+    # shellcheck disable=SC2044
+    for HIT in $(find "$START_DIR" -path "*templates/hooks/pre-commit.sample*" 2>/dev/null); do
```
Comment (rycus86) [inline install.sh]: Nit: this would match `pre-commit.sample.whatever` too. I don't think the trailing `*` hurts the `find` performance too much, but I'd drop it.

File: install.sh [hunk 4]
```diff
@@ -3616,9 +3620,16 @@ install_into_existing_repositories() {
         LOCAL_REPOSITORY_LIST=$(echo "$LOCAL_REPOSITORY_LIST" | sort)
     fi
 
+    TEMP_IFS="
```
Comment (rycus86) [inline install.sh]: Should this go like this?
```
LOOP_IFS="
"
PREVIOUS_IFS="$IFS"
IFS="$LOOP_IFS"
for EXISTING in $LOCAL_REPOSITORY_LIST; do
    unset IFS
    ...
    IFS="$LOOP_IFS"
done
IFS="$PREVIOUS_IFS"
```

File: tests/exec-steps.sh [hunk 5]
```diff
@@ -20,6 +20,8 @@ for STEP in /var/lib/tests/step-*.sh; do
         mkdir -p /usr/share/git-core/templates/hooks
         rm -rf /usr/share/git-core/templates/hooks/*
     fi
+
+    rm -rf ~/test*
```
Comment (rycus86) [inline tests/exec-steps.sh]: Is this related to the whitespace in all the test names? If so, please revert for now.

**Comment by rycus86** [comment]
OK, I think this almost looks ready go. :+1: 
Added some comments for some style changes, plus please revert all the test changes for now.
One (new) whitespace test would be nice, but I'm happy to add one myself after merging this PR.
And I'll also add a new Travis job to automatically change all the paths in the test scripts to have a space, and test the whitespaces in general that way.

**Comment by rycus86** [comment]
Oh `step-006` was failing on Travis with your latest changes. :thinking:

**Comment by gabyx** [comment]
> OK, I think this almost looks ready go. 👍
> Added some comments for some style changes, plus please revert all the test changes for now.
> One (new) whitespace test would be nice, but I'm happy to add one myself after merging this PR.
> And I'll also add a new Travis job to automatically change all the paths in the test scripts to have a space, and test the whitespaces in general that way.

That sounds perfect.
I do the review in moment.
What do you mean by step-006 was failing? I check the tests when I ve done the review. Lets see. Thanks :-)

File: install.sh (hunk 4 above)
Comment (gabyx) [reply install.sh]: I think we should always unset the IFS. not setting it to the previous one.
Meaning when we want to use special behavior, we set the IFS only in the scope we need it, and afterward reset it to `unset IFS` meaning standard wordsplitting is used "<tab><space><newline>"

Comment (rycus86) [reply install.sh]: Ok, sounds good.

File: tests/exec-steps.sh (hunk 5 above)
Comment (gabyx) [reply tests/exec-steps.sh]: No it isn't, these are some artefacts from tests which write to the home directory, as I found :-) 
I 'd leave it :-)

**Comment by gabyx** [comment]
Can we enumerate
```LOOP_IFS_<linenumber>``` and have a pre-commit script which reassigns the line numbers always...?

**Comment by rycus86** [comment]
> Can we enumerate
> `LOOP_IFS_<linenumber>` and have a pre-commit script which reassigns the line numbers always...?

Why is this a concern?

**Comment by gabyx** [comment]
Because of nestet for loops: Using a `LOOP_IFS=...` might  might be reset if another shell function iside the loop resets LOOP_IFS to something else...
Maybe define globals some IFS.
IFS_COMMA_NEWLINE
IFS_NEWLINE
and only use them to reassign IFS
```
IFS="$IFS_COMMA_NEWLINE"
for ...
unset IFS
...
IFS="$IFS_COMMA_NEWLINE"
done

**Comment by rycus86** [comment]
> Because of nestet for loops: Using a `LOOP_IFS=...` might might be reset if another shell function iside the loop resets LOOP_IFS to something else...

OK, what I meant to ask is: is this a real concern here? Does it happen already? How hard/easy it is to make it happen?

If it's not super easy to mess up, I'd not worry about it for now.

**Comment by gabyx** [comment]
See my latest commit. global IFS to use in LOOPS, that should fix also potential future problems.
It think this way its kind of more robust.

File: cli.sh [hunk 6]
```diff
@@ -1064,6 +1072,7 @@ list_shared_hook_repos() {
             SHARED_REPOS_LIST=$(grep -E "^[^#].+$" <"$(pwd)/.githooks/.shared")
 
             echo "$SHARED_REPOS_LIST" | while read -r LIST_ITEM; do
```
Comment (rycus86) [inline cli.sh]: Missing `IFS="$IFS_COMMA_NEWLINE"` before the loop?

File: tests/exec-steps.sh (hunk 5 above)
Comment (rycus86) [reply tests/exec-steps.sh]: OK

**Comment by rycus86** [comment]
Nice, looks good to me! :+1: 
Let's wait for Travis then I'll merge it in.

**Comment by rycus86** [comment]
Hm, somethings seems to have broken around shared repo lists, see steps `082` and `092`.

**Comment by gabyx** [comment]
Seen it :-) Will investigate.

File: cli.sh [hunk 7]
```diff
@@ -66,6 +66,10 @@ set_main_variables() {
     if [ "${CURRENT_GIT_DIR}" = "--git-common-dir" ]; then
         CURRENT_GIT_DIR=".git"
     fi
+
+    # Global IFS for loops
+    IFS_COMMA_NEWLINE="
```
Comment (rycus86) [inline cli.sh]: Oh, I missed this. You say `COMMA_NEWLINE` but doesn't actually include `,`

Comment (gabyx) [reply cli.sh]: damm, thx!

Comment (gabyx) [reply cli.sh]: At least we know that some test fail like that :)

**Comment by rycus86** [comment]
> Hm, somethings seems to have broken around shared repo lists, see steps `082` and `092`.

OK, found the problem I think, left a comment.

**Comment by gabyx** [comment]
Thx!

**Comment by rycus86** [comment]
> Thx!

Thank you! :)
